### PR TITLE
[video_player] synchronize isPlaying state

### PR DESCRIPTION
**Pull request by maRci002**
This was the original PR: flutter/plugins#7198

Resolves flutter/flutter#49081.
Resolves flutter/flutter#120872.

This PR synchronizes `isPlaying` state between native and dart code, for example if a headset is unplugged or a phone call interrupts the video.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by hellohuanlin** [comment]
Feel free to request review again when my comments in the original PR is addressed.

**Comment by maRci002** [comment]
I've addressed your comments but removing those two `dependency_overrides` section results error like this:

> Error (Xcode): ../lib/video_player.dart:441:29: Error: Member not found: 'isPlayingStateUpdate'.

File: packages/video_player/video_player_android/example/pubspec.yaml
Comment (hellohuanlin) [inline packages/video_player/video_player_android/example/pubspec.yaml:37]: remove here

File: packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart:277]: noob question: is `copyWith` functionality enabled by `immutable`?

File: packages/video_player/video_player_avfoundation/example/pubspec.yaml
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/pubspec.yaml:37]: here

**Review by hellohuanlin — APPROVED** [review]
iOS part looks good

**Comment by hellohuanlin** [comment]
> I've addressed your comments but removing those two dependency_overrides section results error like this: Error (Xcode): ../lib/video_player.dart:441:29: Error: Member not found: 'isPlayingStateUpdate'.

My understanding is that what you changed is the standard way of doing things. IIRC I have seen similar practice in other plugins. Without that overrides, the dependency would be the previously published version

File: packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart
Comment (maRci002) [reply packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart:277]: [immutable top-level constant](https://api.flutter.dev/flutter/meta/immutable-constant.html)

> A class is immutable if all of the instance fields of the class, whether defined directly or inherited, are final.

While the `@immutable` annotation is commonly used to indicate that a class is immutable in Dart, it does not automatically provide a `copyWith` method for the class. However, defining a `copyWith` method is a common pattern for creating new instances of immutable objects with modified values. Note that some classes like `Widget` or `StatelessWidget` in Flutter may have the `@immutable` annotation, but they do not necessarily provide a `copyWith` method since they are more like blueprints or configurations rather than data models.

Since `VideoPlayerValue` already using `final` fields it's easy to add `@immutable`.

Since `VideoPlayerController extends ValueNotifier<VideoPlayerValue>` we have to create a new a `value` and since `ValueNotifier` notifies it's listeners when it gets a new `value` it should override `equals` method.

[ValueNotifier<T> class](https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html)

> A [ChangeNotifier](https://api.flutter.dev/flutter/foundation/ChangeNotifier-class.html) that holds a single value.
> 
> When [value](https://api.flutter.dev/flutter/foundation/ValueNotifier/value.html) is replaced with something that is not equal to the old value as evaluated by the equality operator ==, this class notifies its listeners.

**Comment by maRci002** [comment]
How can add back the reviewers?

File: packages/video_player/video_player_web/example/integration_test/video_player_web_test.dart
Comment (ditman) [inline packages/video_player/video_player_web/example/integration_test/video_player_web_test.dart:190]: The timeout of this stream isn't modified (line 180), so this is still waiting for 1 second.

**Review by ditman — APPROVED** [review]
The web bits look good to me, this is a big update for the player, thanks for doing it!

**Review by camsim99 — APPROVED** [review]
The Android parts look good to me!

This PR made me realize the lack of test coverage on the Android native parts, so you _could_ add a test for this change, but definitely not blocking for this PR. More-so a note to self :)

**Comment by maRci002** [comment]
I've added an Android test

**Comment by reidbaker** [comment]
Thank you for your contribution. It looks like the tests are failing. Before we can merge the PR, please see what you can do to resolve the test failures. If you are unsure how to proceed, please reach out for help on the #hackers-new channel.

**Comment by maRci002** [comment]
@reidbaker I fixed the tests.

I still need #3361 PR to be merged according to https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins

File: packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart
Comment (bparrishMines) [inline packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:241]: I think this would be better as `Whether the video is currently playing.`.

File: packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart
Comment (bparrishMines) [inline packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:290]: nit: This should follow

https://dart.dev/guides/language/effective-dart/documentation#do-start-doc-comments-with-a-single-sentence-summary and https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

File: packages/video_player/video_player/lib/video_player.dart
Comment (bparrishMines) [inline packages/video_player/video_player/lib/video_player.dart:441]: This change will need a test.

**Review by bparrishMines — CHANGES_REQUESTED** [review]

**Review by bparrishMines — APPROVED** [review]
LGTM

It looks like this also already has approval for, iOS, web, and Android. So, I'll take a look at https://github.com/flutter/packages/pull/3361.

**Comment by BraveEvidence** [comment]
This will help https://www.youtube.com/watch?v=IMQdSTlTXjA

**Comment by maRci002** [comment]
In my opinion this PR is ready

- removed `dependency_overrides`
- removed `// ignore: missing_enum_constant_in_switch` which was created by #3435
- updated `video_player_platform_interface` dependencies to `6.1.0` since #3361 PR increased the version

**Comment by gmackall** [comment]
Is this PR missing anything in order to be able to merge?

**Comment by tarrinneal** [comment]
> Is this PR missing anything in order to be able to merge?

Seems ready to me.

**Review by tarrinneal — APPROVED** [review]

**Pull request merged**
